File: sdks/js/test/drift.test.ts
/**
 * The drift test.
 *
 * hostd's JSON tags are the contract (`internal/api/types.go:1-7`). This SDK
 * keeps a second copy of them, which is exactly the situation where a copy
 * rots silently: a field added to `Machine` in Go would simply never reach a
 * consumer, and nothing would be red.
 *
 * So the copy is checked against its source on every `npm test`: parse every
 * non-test Go file in the two packages that own wire shapes, parse
 * `src/types.ts`, and fail naming the struct and the tag when the two
 * disagree in either direction.
 */

import test from 'node:test'
import assert from 'node:assert/strict'
import { readFileSync, readdirSync, existsSync } from 'node:fs'
import { join } from 'node:path'

const repoRoot = join(import.meta.dirname, '..', '..', '..')
const apiDir = join(repoRoot, 'apps', 'hostd', 'internal', 'api')
// `internal/compose` arrives with #30. Until then there is nothing to walk;
// the day it lands, its structs are checked without touching this file.
const composeDir = join(repoRoot, 'apps', 'hostd', 'internal', 'compose')
const typesTS = join(import.meta.dirname, '..', 'src', 'types.ts')

/** Strips Go/TS line and block comments, so a comment cannot fake a field. */
function stripComments(src: string): string {
  return src.replace(/\/\*[\s\S]*?\*\//g, '').replace(/\/\/.*$/gm, '')
}

interface GoStruct {
  name: string
  file: string
  tags: string[]
}

/**
 * Every `type X struct { ... }` in a directory, with its JSON tag names.
 *
 * gofmt puts the closing brace of a top-level struct at column 0, which is
 * what makes the block boundary reliable without a real parser.
 */
function parseGoStructs(dir: string, prefix = ''): GoStruct[] {
  const out: GoStruct[] = []
  for (const file of readdirSync(dir).sort()) {
    if (!file.endsWith('.go') || file.endsWith('_test.go')) continue
    const lines = stripComments(readFileSync(join(dir, file), 'utf8')).split('\n')
    for (let i = 0; i < lines.length; i++) {
      const open = /^type ([A-Za-z0-9_]+) struct \{/.exec(lines[i]!)
      if (!open) continue
      const tags: string[] = []
      for (i++; i < lines.length && lines[i] !== '}'; i++) {
        const tag = /json:"([^"]*)"/.exec(lines[i]!)
        if (!tag) continue
        const name = tag[1]!.split(',')[0]!
        if (name && name !== '-') tags.push(name)
      }
      // A struct with no tags is not a wire shape (api.Deps, for one).
      if (tags.length > 0) out.push({ name: prefix + open[1]!, file, tags })
    }
  }
  return out
}

/** Every `export interface X { ... }` in types.ts, with its property names. */
function parseTSInterfaces(file: string): Map<string, string[]> {
  const lines = stripComments(readFileSync(file, 'utf8')).split('\n')
  const out = new Map<string, string[]>()
  for (let i = 0; i < lines.length; i++) {
    const open = /^export interface ([A-Za-z0-9_]+) \{/.exec(lines[i]!)
    if (!open) continue
    const props: string[] = []
    for (i++; i < lines.length && lines[i] !== '}'; i++) {
      const prop = /^\s+([A-Za-z0-9_]+)\??\s*:/.exec(lines[i]!)
      if (prop) props.push(prop[1]!)
    }
    out.set(open[1]!, props)
  }
  return out
}

const goStructs = [
  ...parseGoStructs(apiDir),
  ...(existsSync(composeDir) ? parseGoStructs(composeDir, 'Compose') : []),
]
const tsInterfaces = parseTSInterfaces(typesTS)

test('the parsers found something to compare', () => {
  // A moved or renamed directory must fail loudly rather than pass vacuously
  // by finding nothing on either side.
  assert.ok(
    goStructs.length >= 20,
    `only ${goStructs.length} tagged structs found under ${apiDir}; the directory moved?`,
  )
  assert.ok(tsInterfaces.size >= 20, `only ${tsInterfaces.size} interfaces found in src/types.ts`)
})

test('src/types.ts mirrors every wire struct hostd serves', () => {
  const problems: string[] = []
  for (const s of goStructs) {
    const props = tsInterfaces.get(s.name)
    if (!props) {
      problems.push(`types.ts has no interface ${s.name} (hostd ${s.file}: ${s.tags.join(', ')})`)
      continue
    }
    const missing = s.tags.filter((t) => !props.includes(t))
    const extra = props.filter((p) => !s.tags.includes(p))
    if (missing.length || extra.length) {
      problems.push(
        `types.ts ${s.name} is missing [${missing.join(', ')}] / ` +
          `carries extra [${extra.join(', ')}] (hostd ${s.file})`,
      )
    }
  }
  assert.deepEqual(problems, [])
})

test('the frame constants match hostd', () => {
  const src = readFileSync(join(apiDir, 'types.go'), 'utf8')
  const ts = readFileSync(typesTS, 'utf8')
  const frames = [...src.matchAll(/^\s*(Frame[A-Za-z]+)\s+byte\s*=\s*(\d+)$/gm)]
  assert.ok(frames.length >= 3, 'no Frame constants found in hostd types.go')
  for (const [, name, value] of frames) {
    const mine = new RegExp(`^export const ${name} = (\\d+)$`, 'm').exec(ts)
    assert.ok(mine, `types.ts does not export ${name} (hostd types.go has it = ${value})`)
    assert.equal(mine[1], value, `types.ts ${name} is ${mine[1]}, hostd says ${value}`)
  }
})

File: sdks/elixir/test/drift_test.exs
defmodule Pilots.DriftTest do
  @moduledoc """
  The drift test, in the shape this client needs one.

  The other three SDKs mirror hostd's structs as types and compare field for
  field. This one returns plain maps with the server's own keys, so there is
  no second copy of a field name to rot. What CAN rot is the ROUTE TABLE: a
  function here names a path as a string, and a path hostd renamed or removed
  would fail only at runtime, against a real fleet, for whoever ran it first.

  So the check is that every path this module spells out is a path hostd's own
  route table registers. It parses `internal/api/routes.go`, which is the file
  that decides what exists.
  """
  use ExUnit.Case, async: true

  @api_dir Path.expand("../../../apps/hostd/internal/api", __DIR__)

  defp hostd_routes do
    routes_go = Path.join(@api_dir, "routes.go")

    routes_go
    |> File.read!()
    |> then(&Regex.scan(~r/mux\.Handle(?:Func)?\("(?:([A-Z]+) )?([^"]+)"/, &1))
    |> Enum.map(fn
      [_, _method, path] -> path
    end)
    |> Enum.map(&normalise/1)
    |> MapSet.new()
  end

  # `{id}` and `{name}` are one shape as far as this test is concerned: what
  # it checks is that the SEGMENTS around them are the ones hostd serves.
  defp normalise(path), do: Regex.replace(~r/\{[a-zA-Z]+\}/, path, "{}")

  defp client_paths do
    lib = Path.expand("../lib", __DIR__)

    for file <- Path.wildcard(Path.join(lib, "**/*.ex")),
        line <- File.read!(file) |> String.split("\n"),
        # A path in this client is always a string starting `/v1/` (or `/mcp`),
        # interpolated with `#{...}` where hostd has a `{param}`.
        capture <- Regex.scan(~r|"(/v1/[^"]*)"|, line),
        [_, path] = capture,
        into: MapSet.new() do
      path
      |> String.replace(~r/\#\{[^}]*\}/, "{}")
      |> normalise()
    end
  end

  test "hostd's route table was found" do
    routes = hostd_routes()

    # A moved file must fail loudly rather than pass vacuously by finding
    # nothing on either side.
    assert MapSet.size(routes) >= 20,
           "only #{MapSet.size(routes)} routes parsed from #{@api_dir}; the file moved?"

    assert MapSet.member?(routes, "/v1/machines")
  end

  test "every path this client calls is one hostd serves" do
    routes = hostd_routes()
    mine = client_paths()

    assert MapSet.size(mine) >= 15,
           "only #{MapSet.size(mine)} paths found in lib/; the parse broke?"

    unknown = MapSet.difference(mine, routes) |> MapSet.to_list() |> Enum.sort()

    assert unknown == [],
           """
           These paths are not in hostd's route table:

             #{Enum.join(unknown, "\n  ")}

           Either the route was renamed and this client was not, or the path
           was typed wrong. hostd's table is apps/hostd/internal/api/routes.go.
           """
  end

  test "the routes a client of this shape needs are all reachable" do
    mine = client_paths()

    # Not every hostd route belongs in every client, but these are the ones a
    # caller of THIS module can reach, and a silent removal of one would be a
    # feature quietly disappearing from the Elixir surface.
    for path <- [
          "/v1/machines",
          "/v1/machines/{}",
          "/v1/machines/{}/exec",
          "/v1/machines/{}/checkpoints",
          "/v1/machines/{}/promote",
          "/v1/checkpoints/{}/restore",
          "/v1/services",
          "/v1/services/{}/deploy",
          "/v1/builds",
          "/v1/plan",
          "/v1/whoami",
          "/v1/health"
        ] do
      assert MapSet.member?(mine, path), "the client no longer calls #{path}"
    end
  end
end
